File: day16_part1_2016.dart

import 'dart:io';

void main() {
  String input = File('input.txt').readAsStringSync().trim();
  
  while (input.length < 272) {
    String a = input;
    String b = a.split('').reversed.map((c) => c == '0' ? '1' : '0').join('');
    input = '$a' '0' '$b';
  }
  
  String data = input.substring(0, 272);
  
  String checksum = generateChecksum(data);
  
  print(checksum);
}

String generateChecksum(String data) {
  String checksum = '';
  
  do {
    checksum = '';
    for (int i = 0; i < data.length - 1; i += 2) {
      if (data[i] == data[i + 1]) {
        checksum += '1';
      } else {
        checksum += '0';
      }
    }
    
    data = checksum;
  } while (checksum.length % 2 == 0);
  
  return checksum;
}

File: objectivec/day16_part1_2016.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSError *error;
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
        
        NSString *data = input;
        while (data.length < 272) {
            NSString *copy = [data copy];
            NSMutableString *reversed = [NSMutableString stringWithString:@""];
            for (int i = (int)copy.length - 1; i >= 0; i--) {
                if ([copy characterAtIndex:i] == '0') {
                    [reversed appendString:@"1"];
                } else {
                    [reversed appendString:@"0"];
                }
            }
            data = [NSString stringWithFormat:@"%@0%@", data, reversed];
        }
        
        data = [data substringToIndex:272];
        
        NSMutableString *checksum = [NSMutableString stringWithString:data];
        while (checksum.length % 2 == 0) {
            NSMutableString *newChecksum = [NSMutableString stringWithString:@""];
            for (int i = 0; i < checksum.length; i += 2) {
                if ([checksum characterAtIndex:i] == [checksum characterAtIndex:i + 1]) {
                    [newChecksum appendString:@"1"];
                } else {
                    [newChecksum appendString:@"0"];
                }
            }
            checksum = newChecksum;
        }
        
        printf("%s\n", [checksum UTF8String]);
    }
    return 0;
}